### ISSUE TITLE
Consider using Dagre.NET to visualize and perhaps edit/export the transformer architecture

### ISSUE THREAD

**Issue by GeorgeS2019** [zhongkaifu/Seq2SeqSharp#32]
**Is your feature request related to a problem? Please describe.**
We need a way to **visualize/edit/export (onnx) transformer architecture for .NET community.**

**Describe the solution you'd like**
[Dagre.NET and Dendrite could evolve to support this need](https://github.com/fel88/Dagre.NET/issues/2#issuecomment-929501734). Seq2SeqSharp could export the architecture in Json format for import into Dendrite or use Dagre.NET to export images.

**Comment by zhongkaifu** [comment]
Thanks @GeorgeS2019 . It's a great idea.
Did you know if Dagre.NET project has any example of Json format for models ? If so, I can take a look. I'm also thinking about getting rid of BinaryFormatter for model serialization and de-serialization. :)


Thanks
Zhongkai Fu

**Comment by GeorgeS2019** [comment]
@zhongkaifu 

It adhere as close as possible to the [JavaScript library version](https://github.com/dagrejs/dagre/wiki)

if there is anything you need, please feedback directly [here](https://github.com/fel88/Dagre.NET/issues/2).

# Csharp versoin
```cs
DagreInputGraph dg = new DagreInputGraph();
var nd1 = dg.AddNode();
var nd2 = dg.AddNode();
dg.AddEdge(nd1, nd2);
dg.Layout(); 
Console.WriteLine($"node1 : {nd1.X} {nd1.Y}");
Console.WriteLine($"node2 : {nd2.X} {nd2.Y}");
```

# JavaScript version
```js
// Create a new directed graph 
var g = new dagre.graphlib.Graph();

// Set an object for the graph label
g.setGraph({});

// Default to assigning a new object as a label for each new edge.
g.setDefaultEdgeLabel(function() { return {}; });

// Add nodes to the graph. The first argument is the node id. The second is
// metadata about the node. In this case we're going to add labels to each of
// our nodes.
g.setNode("kspacey",    { label: "Kevin Spacey",  width: 144, height: 100 });
g.setNode("swilliams",  { label: "Saul Williams", width: 160, height: 100 });
g.setNode("bpitt",      { label: "Brad Pitt",     width: 108, height: 100 });
g.setNode("hford",      { label: "Harrison Ford", width: 168, height: 100 });
g.setNode("lwilson",    { label: "Luke Wilson",   width: 144, height: 100 });
g.setNode("kbacon",     { label: "Kevin Bacon",   width: 121, height: 100 });

// Add edges to the graph.
g.setEdge("kspacey",   "swilliams");
g.setEdge("swilliams", "kbacon");
g.setEdge("bpitt",     "kbacon");
g.setEdge("hford",     "lwilson");
g.setEdge("lwilson",   "kbacon");

dagre.layout(g);
```

**Comment by fel88** [comment]
> Did you know if Dagre.NET project has any example of Json format for models ? 

Dendrite supports only ONNX so far, but I can implement any Json format that you provide.

**Comment by zhongkaifu** [comment]
Thanks @GeorgeS2019  and @fel88 . It's really helpful.

In Seq2SeqSharp, I used to use Microsoft.Msagl.Drawing to draw operators/layers/networks, however, since it doesn't support .NET core and .NET 5.0, so I comment it out. But I still keep the empty of it. If you look at VisualizeNodes method in https://github.com/zhongkaifu/Seq2SeqSharp/blob/master/Seq2SeqSharp/Tools/ComputeGraphTensor.cs file, you will find these code.

So, one thing we could do it to replace those old code I comment out to your code in Dagre.NET project. In addition, methods in ComputeGraphTensor.cs are operator level, so they are good entry pointers for visualization and ONNX export.

I'm glad if you could do it and let me know if you have any further question on it.

Thanks
Zhongkai Fu

**Comment by GeorgeS2019** [comment]
@zhongkaifu  @fel88 

Is there a graph sub patterns or regex that can be used to search the complete  e.g [BERT.ONNX graph ](https://github.com/fel88/Dagre.NET/issues/2#issuecomment-930173945)to create groups or nested graphs to create boundaries around e.g. Encoder, Decoder of a transformer architecture?

**Comment by zhongkaifu** [comment]
@GeorgeS2019  Yes, Seq2SeqSharp does have logic to create boundaries for sub-graphs. You can check method "IComputeGraph CreateSubGraph(string name)" in https://github.com/zhongkaifu/Seq2SeqSharp/blob/master/Seq2SeqSharp/Tools/ComputeGraphTensor.cs file as well. However, I  already commented it out due to the same reason in above.

**Comment by GeorgeS2019** [comment]
@zhongkaifu  thnx for the valuable tip :-)
